lib/espi_dni/token_supervisor.ex

```diff
@@ -0,0 +1,72 @@
+defmodule EspiDni.TokenSupervisor do
+  use GenServer
+  use Supervisor
+  use Timex
+  require Logger
+  alias EspiDni.Team
+  alias EspiDni.Repo
+  import Ecto.Query, only: [from: 1, from: 2]
+
+  @expiring_soon_in_seconds 60
+  @refresh_wait_in_seconds 2_700
+  @milliseconds_in_seconds 1_000
+
+  def start_link do
+    GenServer.start_link(__MODULE__, :ok, [name: __MODULE__])
+  end
+
+  def init(:ok) do
+    for team <- teams_with_refresh_tokens do
+      queue_for_refresh(team)
+    end
+    {:ok, %{}}
+  end
+
+  def queue_for_refresh(team) do
+    diff = expires_in_unix(team) - :os.system_time(:seconds)
+
+    if diff < @expiring_soon_in_seconds do
+      # just do it immediately if it will expire soon
+      Task.async fn ->
+        __MODULE__.refresh_token!(team)
+      end
+    else
+      # otherwise queue it for later
+      queue_delay = queue_delay_ms(diff)
+      Logger.info "Queueing token refresh for team: #{team.id} in #{queue_delay}ms"
+      :timer.apply_after(queue_delay, __MODULE__, :refresh_token!, [team])
+    end
+  end
```

Comment:
I think, you could also move the tow blocks in if and else into 2 private methods respectively and give them proper names (e.g. `refresh_now` and `queue_for_later`), so you wouldn't need the comments.